File: src/main/java/cube8540/oauth/authentication/credentials/oauth/client/provider/ClientCredentialsToken.java
package cube8540.oauth.authentication.credentials.oauth.client.provider;

import cube8540.oauth.authentication.credentials.oauth.OAuth2ClientDetails;
import org.springframework.security.authentication.AbstractAuthenticationToken;
import org.springframework.security.core.GrantedAuthority;

import java.util.Collection;

public class ClientCredentialsToken extends AbstractAuthenticationToken {

    private Object principal;
    private Object credentials;

    public ClientCredentialsToken(Object principal, Object credentials) {
        super(null);
        this.principal = principal;
        this.credentials = credentials;
        super.setAuthenticated(false);
    }

    protected ClientCredentialsToken(Object principal, Object credentials, Collection<? extends GrantedAuthority> authorities) {
        super(authorities);
        this.principal = principal;
        this.credentials = credentials;
        super.setAuthenticated(true);
    }

    @Override
    public Object getCredentials() {
        return credentials;
    }

    @Override
    public Object getPrincipal() {
        return principal;
    }

    @Override
    public String getName() {
        if (principal instanceof String) {
            return principal.toString();
        } else if (principal instanceof OAuth2ClientDetails) {
            return ((OAuth2ClientDetails) principal).clientId();
        } else {
            return super.getName();
        }
    }

    @Override
    public void setAuthenticated(boolean authenticated) {
        throw new IllegalArgumentException("this operation is not supported");
    }

    @Override
    public void eraseCredentials() {
        super.eraseCredentials();
        this.credentials = null;
    }
}

File: src/main/kotlin/cube8540/oauth/authentication/credentials/oauth/security/provider/ClientCredentialsToken.kt
package cube8540.oauth.authentication.credentials.oauth.security.provider

import cube8540.oauth.authentication.credentials.oauth.security.OAuth2ClientDetails
import org.springframework.security.authentication.AbstractAuthenticationToken
import org.springframework.security.core.GrantedAuthority
import java.util.*

class ClientCredentialsToken(
    private val principal: Any?,

    private var credentials: Any?,

    authorities: Collection<GrantedAuthority>?
): AbstractAuthenticationToken(authorities) {

    init {
        if (authorities != null) {
            super.setAuthenticated(true)
        } else {
            super.setAuthenticated(false)
        }
    }

    constructor(principal: Any?, credentials: Any?): this(principal, credentials, null)

    override fun getPrincipal(): Any? = principal

    override fun getCredentials(): Any? = credentials

    override fun getName(): String = when (principal) {
        is String -> principal
        is OAuth2ClientDetails -> principal.clientId
        else -> super.getName()
    }

    override fun setAuthenticated(authenticated: Boolean) {
        throw IllegalArgumentException("this operation is not supported")
    }

    override fun eraseCredentials() {
        super.eraseCredentials()
        this.credentials = null
    }

    override fun equals(other: Any?): Boolean = when {
        other == null -> false
        other is ClientCredentialsToken -> {
            other.principal == this.principal &&
                    other.credentials == this.credentials &&
                    other.authorities == this.authorities
        }
        else -> false
    }

    override fun hashCode(): Int = Objects.hash(principal, credentials, authorities)
}